src/core/event-handler.js

```diff
@@ -178,67 +218,55 @@ class EventHandler {
      * obj.fire('test', 'This is the message');
      */
     fire(name, arg1, arg2, arg3, arg4, arg5, arg6, arg7, arg8) {
-        if (!name || !this._callbacks[name])
+        if (!name)
             return this;
 
-        let callbacks;
+        const callbacksInitial = this._callbacks.get(name);
+        if (!callbacksInitial)
+            return this;
 
-        if (!this._callbackActive[name]) {
-            this._callbackActive[name] = this._callbacks[name];
-        } else {
-            if (this._callbackActive[name] === this._callbacks[name])
-                this._callbackActive[name] = this._callbackActive[name].slice();
+        let callbacks;
 
-            callbacks = this._callbacks[name].slice();
+        if (!this._callbackActive.has(name)) {
+            // when starting callbacks execution ensure we store a list of initial callbacks
+            this._callbackActive.set(name, callbacksInitial);
+        } else if (this._callbackActive.get(name) !== callbacksInitial) {
+            // if we are trying to execute a callback while there is an active execution right now
+            // and the active list has been already modified,
+            // then we go to an unoptimized path and clone callbacks list to ensure execution consistency
+            callbacks = Array.from(callbacksInitial);
         }
 
-        // TODO: What does callbacks do here?
-        // In particular this condition check looks wrong: (i < (callbacks || this._callbackActive[name]).length)
-        // Because callbacks is not an integer
+        // @ts-ignore: Array is possibly 'undefined'
```

Comment:
We don't use `@ts-ignore` anywhere across the codebase. I know it's nice to suppress the errors VS Code shows, but for now, we haven't taken a decision to actively use `@ts-ignore`. If we did, there'd be probably many hundreds of them all over the source.